torch/_inductor/fx_passes/post_grad.py

```diff
@@ -802,3 +805,162 @@ def repl(inp, x1, x2):
+            continue
+
+        if not node.kwargs.get("device") == torch.device("cpu"):
+            continue
+
+        constructors.append(node)
+
+    # not handling multiple cuda devices initially
+    if not constructors or len(cuda_devices) != 1:
+        return
+
+    movable_constructors = find_movable_constructors(graph, constructors)
+
+    for node in movable_constructors:
+        kwargs = node.kwargs.copy()
+        kwargs["device"] = next(iter(cuda_devices))
+        node.kwargs = kwargs
+
+
+def find_movable_constructors(graph, constructors: List[fx.Node]) -> Set[fx.Node]:
+    """
+    Starting from the cpu constructors, iterate through the graph and test that all of their
+    downstream uses can safely be moved to cpu.
+    """
+    cpu_indeg: Dict[fx.Node, int] = get_cpu_indeg_count(graph)
+
+    # which constructors cannot be moved to cuda
+    cannot_move_to_cuda: Set[fx.Node] = set()
+
+    # For any node in the graph, which constructors does it have a dependency on
+    constructor_dependencies: Dict[fx.Node, Set[fx.Node]] = defaultdict(set)
+
+    # if a cpu node has a dependency on two different cpu constructors,
+    # then if either constructor cannot be moved to cuda, the other cannot as well.
+    # In this case any node with a dependency on one will have a dependency on the other
+    equal_constructor_sets: Dict[fx.Node, Set[fx.Node]] = {c: {c} for c in constructors}
+
+    def make_dependencies_equivalent(
+        set1: Set[fx.Node], set2: Set[fx.Node]
+    ) -> Set[fx.Node]:
+        # could use union find but not worth complexity here
+        set1.update(set2)
+        for obj in set1:
+            equal_constructor_sets[obj] = set1
+        return set1
+
+    queue: List[fx.Node] = list(constructors)
+
+    for c in queue:
+        constructor_dependencies[c].add(c)
+
+    while queue:
+        node = queue.pop()
+        dependencies = constructor_dependencies[node]
+
+        for user in node.users:
+            if cannot_be_moved_to_cuda(user):
+                cannot_move_to_cuda.update(dependencies)
+                break
+
+            # this node was used on a op which takes in multiple devices and output a cuda
+            # tensor. we can convert its cpu input to cuda without making further changes
+            node_device = get_node_device(user)
+            if (
+                allows_mixed_devices(user.target)
+                and node_device
+                and node_device.type == "cuda"
+            ):
+                del cpu_indeg[user]
+            else:
+                # otherwise, we should continue look at its downstream uses
+                cpu_indeg[user] -= 1
+                if cpu_indeg[user] == 0:
+                    del cpu_indeg[user]
+                    queue.append(user)
+
+            unioned_set = make_dependencies_equivalent(
+                dependencies, constructor_dependencies[user]
+            )
+            constructor_dependencies[user] = unioned_set
```

Comment:
it's weird to me that you didn't fold this into the function

Reply:
kind of leaves it open for a more efficient impl (union find) if we needed it.. but yea fair